plugins/uf2/fu-uf2-device.c

```diff
@@ -0,0 +1,390 @@
+/*
+ * Copyright (C) 2021 Richard Hughes <richard@hughsie.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#include "config.h"
+
+#include <fwupdplugin.h>
+
+#include "fu-uf2-device.h"
+#include "fu-uf2-firmware.h"
+
+struct _FuUf2Device {
+	FuUdevDevice parent_instance;
+	guint64 family_id;
+};
+
+G_DEFINE_TYPE(FuUf2Device, fu_uf2_device, FU_TYPE_UDEV_DEVICE)
+
+static FuFirmware *
+fu_uf2_device_prepare_firmware(FuDevice *device,
+			       GBytes *fw,
+			       FwupdInstallFlags flags,
+			       GError **error)
+{
+	FuUf2Device *self = FU_UF2_DEVICE(device);
+	g_autoptr(FuFirmware) firmware = fu_uf2_firmware_new();
+
+	if (!fu_firmware_parse(firmware, fw, flags, error))
+		return NULL;
+
+	/* check the family_id matches if we can read the old firmware */
+	if (self->family_id > 0 && fu_firmware_get_idx(firmware) > 0 &&
+	    self->family_id != fu_firmware_get_idx(firmware)) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_NOT_SUPPORTED,
+			    "family ID was different, expected 0x%08x bytes and got 0x%08x",
+			    (guint)self->family_id,
+			    (guint)fu_firmware_get_idx(firmware));
+		return NULL;
+	}
+
+	/* success: but return the raw data */
+	return fu_firmware_new_from_bytes(fw);
+}
+
+static gboolean
+fu_uf2_device_probe_current_fw(FuDevice *device, GBytes *fw, GError **error)
+{
+	g_autofree gchar *csum_sha256 = NULL;
+	g_autoptr(FuFirmware) firmware = fu_uf2_firmware_new();
+	g_autoptr(GBytes) fw_raw = NULL;
+
+	/* parse to get version */
+	if (!fu_firmware_parse(firmware, fw, FWUPD_INSTALL_FLAG_NONE, error))
+		return FALSE;
+	if (fu_firmware_get_version(firmware) != NULL)
+		fu_device_set_version(device, fu_firmware_get_version(firmware));
+
+	/* add instance ID for quirks */
+	if (fu_firmware_get_idx(firmware) != 0x0) {
+		g_autofree gchar *id0 = NULL;
+		id0 = g_strdup_printf("UF2\\FAMILY_%08X", (guint32)fu_firmware_get_idx(firmware));
```

Comment:
since family is just an integer - how would you feel about adding a VENDOR_%s as well into the string?

Reply:
In theory, the family ID is like a registry -- so the vendor is not required.

Reply:
You also don't get the vendor ID from the firmware, but you do get the family ID.

Reply:
what is to stop cheap vendor XYZ from using reference design and forgetting to change it?

Reply:
the family ID is specified in the firmware, and it's the only thing we've got to match up the custom firmware with the generic devboard.

Reply:
I've now tested this on the *Adafruit Trinket M0* and *Adafruit Trinket M0* and it seems to work well.